File: backend/src/services/ocr_service.py
import sys
import json
import easyocr
import numpy as np
from PIL import Image

def process_image(image_path):
    try:
        print("Initializing EasyOCR...", file=sys.stderr)
        reader = easyocr.Reader(['en'], model_storage_directory='/app/models')
        
        print("Loading image...", file=sys.stderr)
        image = Image.open(image_path)
        
        print("Processing image with OCR...", file=sys.stderr)
        result = reader.readtext(image_path)
        
        # Convert numpy arrays to lists for JSON serialization
        processed_result = []
        for bbox, text, conf in result:
            processed_result.append({
                'bbox': [[float(x) for x in point] for point in bbox],
                'text': text,
                'confidence': float(conf)
            })
        
        print(json.dumps({
            'success': True,
            'data': processed_result
        }))
        
    except Exception as e:
        print(f"Error processing image: {str(e)}", file=sys.stderr)
        print(json.dumps({
            'success': False,
            'error': str(e)
        }))
        sys.exit(1)

if __name__ == "__main__":
    if len(sys.argv) != 2:
        print(json.dumps({
            'success': False,
            'error': 'Image path not provided'
        }))
        sys.exit(1)
        
    image_path = sys.argv[1]
    process_image(image_path)

File: backend/src/services/ocrService.js
const { createWorker } = require('tesseract.js');

class OCRService {
  constructor() {
    this.workers = {};
  }

  async initWorker(language) {
    if (!this.workers[language]) {
      const worker = await createWorker();
      await worker.loadLanguage(language);
      await worker.initialize(language);
      this.workers[language] = worker;
    }
    return this.workers[language];
  }

  async performOCR(imagePath, language) {
    try {
      if (!['eng', 'hin', 'mar'].includes(language)) {
        throw new Error('Invalid language. Use "eng", "hin", or "mar".');
      }

      const worker = await this.initWorker(language);
      console.log(`Processing image in ${language}:`, imagePath);
      
      const result = await worker.recognize(imagePath);
      console.log('OCR completed');

      const words = result.data.words.map(word => ({
        text: word.text,
        confidence: word.confidence,
        bbox: [
          [word.bbox.x0, word.bbox.y0],
          [word.bbox.x1, word.bbox.y0],
          [word.bbox.x1, word.bbox.y1],
          [word.bbox.x0, word.bbox.y1]
        ]
      }));

      return words;
    } catch (error) {
      console.error('OCR Error:', error);
      throw error;
    }
  }
}

module.exports = new OCRService();
